ios/sound.mm

```diff
@@ -24,209 +24,323 @@
-{
-  // get the driver: check if devices are capable
-  // reload the driver list of available sound devices
-  //GetAvailableInOutDevices();
-
-  // find driver index from given driver name
-  //int iDriverIdx = INVALID_INDEX; // initialize with an invalid index
-
-  /*for ( int i = 0; i < MAX_NUMBER_SOUND_CARDS; i++ )
-  {
-      if ( strDriverName.compare ( strDriverNames[i] ) == 0 )
-      {
-          iDriverIdx = i;
-      }
-  }
-
-  // if the selected driver was not found, return an error message
-  if ( iDriverIdx == INVALID_INDEX )
-  {
-      return tr ( "The current selected audio device is no longer present in the system." );
-  }
-
-  // check device capabilities if it fulfills our requirements
-  const QString strStat = CheckDeviceCapabilities ( iDriverIdx );
-
-  // check if device is capable and if not the same device is used
-  if ( strStat.isEmpty() && ( strCurDevName.compare ( strDriverNames[iDriverIdx] ) != 0 ) )
-  {
-  }
-  // set the left/right in/output channels
-   
-  return strStat;
-   */
-    return "";
-}
-
-QString CSound::CheckDeviceCapabilities ( const int iDriverIdx )
+// TODO - CSound::Init is called multiple times at launch to verify device capabilities.
+// For iOS though, Init takes long, so should better reduce those inits for iOS (Now: 9 times/launch).
+int CSound::Init ( const int iCoreAudioBufferSizeMono )
 {
-  /*// This function checks if our required input/output channel
-  // properties are supported by the selected device. If the return
-  // string is empty, the device can be used, otherwise the error
-  // message is returned.
-
-  // check sample rate of input
-  audioSession.setPreferredSampleRate( SYSTEM_SAMPLE_RATE_HZ );
-  // if false, try to set it
-  if ( audioSession.sampleRate != SYSTEM_SAMPLE_RATE_HZ )
-  {
-    throw CGenErr ( tr ("Could not set sample rate. Please close other applications playing audio.") );
-  }
-
-  // special case with 4 input channels: support adding channels
-    if ( ( lNumInChan == 4 ) && bInputChMixingSupported )
+    try
     {
-        // add four mixed channels (i.e. 4 normal, 4 mixed channels)
-        lNumInChanPlusAddChan = 8;
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
+        this->iCoreAudioBufferSizeMono = iCoreAudioBufferSizeMono;
+
+        // set internal buffer size value and calculate stereo buffer size
+        iCoreAudioBufferSizeStereo = 2 * iCoreAudioBufferSizeMono;
+
+        // create memory for intermediate audio buffer
+        vecsTmpAudioSndCrdStereo.Init ( iCoreAudioBufferSizeStereo );
+
+        AVAudioSession* sessionInstance = [AVAudioSession sharedInstance];
+
+        // we are going to play and record so we pick that category
+        NSError* error = nil;
+        [sessionInstance setCategory:AVAudioSessionCategoryPlayAndRecord
+                         withOptions:AVAudioSessionCategoryOptionMixWithOthers | AVAudioSessionCategoryOptionDefaultToSpeaker |
```

Comment:
Do we really want these options? I would have thought to use jamulus we'd want exclusive and not over bluetooth or airplay due to latency concerns.

Reply:
In case you want to join as Listener, delay is not an issue.
Bluetooth (and probably AirPlay too) is not working properly though because of sample rate difference.